File: w3/dz_cpp/bcw4/bcw4/Unit/Unit.cpp
#include <iostream>
#include "Unit.h"

void Unit::ensureIsAlive() {
    if ( this->hitPoints == 0 ) {
        throw UnitIsDead();
    }
}

Unit::Unit(const std::string& name, int hp, int dmg) {
    this->damage = dmg;
    this->hitPoints = hp;
    this->hitPointsLimit = hp;
    this->name = name;
}

Unit::~Unit() {}

int Unit::getDamage() const {
    return this->damage;
}

int Unit::getHitPoints() const {
    return this->hitPoints;
}

int Unit::getHitPointsLimit() const {
    return this->hitPointsLimit;
}

const std::string& Unit::getName() const {
    return this->name;
}

void Unit::addHitPoints(int hp) {
    this->ensureIsAlive();
    std::cout << this->name << " restored health!" << "\n" << std::endl;
    
    if( hp > this->hitPointsLimit - this->hitPoints ) {
        this->hitPoints = this->hitPointsLimit;
        return;
    }
    this->hitPoints += hp;
}

void Unit::takeDamage(int dmg) {
    this->ensureIsAlive();

    if ( dmg >= this->hitPoints ) {
        this->hitPoints = 0;
        return;
    }
    this->hitPoints -= dmg;
}

void Unit::attack(Unit& enemy) {
    enemy.takeDamage(this->damage);
    std::cout << this->name << " attacked " << enemy.name << "\n" << std::endl;

    enemy.counterAttack(*this);
}

void Unit::counterAttack(Unit& enemy) {
   enemy.takeDamage(this->damage / 2);
   std::cout << this->name << " counterattacked " << enemy.name << "\n" << std::endl;
}

 std::ostream& operator<<(std::ostream& out, const Unit& unit) {
    if ( unit.getHitPoints() == 0 ) {
        return out << "Unit '" << unit.getName() << "' is dead!"; 
    }
    out << "Unit: " << unit.getName() << "\n";
    out << "Health: " << unit.getHitPoints() << "/" << unit.getHitPointsLimit() << "HP\n";
    out << "Damage: " << unit.getDamage() << "dmg\n";
    return out;
 }

File: w3/dz_java/src/Unit/Unit.java
public class Unit {
    private int damage;
    private int hitPoints;
    private int hitPointsLimit;
    private String name;

    private void ensureIsAlive() throws UnitIsDead {
        if ( this.hitPoints == 0 ) {
            throw new UnitIsDead();
        }
    }

    public Unit(String name, int hp, int dmg) {
        this.damage = dmg;
        this.hitPoints = hp;
        this.hitPointsLimit = hp;
        this.name = name;
    }

    public int getDamage() {
        return damage;
    }

    public int getHitPoints() {
        return hitPoints;
    }

    public int getHitPointsLimit() {
        return hitPointsLimit;
    }

    public String getName() {
        return name;
    }

    public void addHitPoints(int hp) throws UnitIsDead {
        this.ensureIsAlive();

        System.out.println(this.name + " restored heath!\n");

        if ( hp > this.hitPointsLimit - this.hitPoints ) {
            this.hitPoints = this.hitPointsLimit;
            return;
        }
        this.hitPoints += hp;
    }

    public void takeDamage(int dmg) throws UnitIsDead {
        this.ensureIsAlive();

        if ( dmg > this.hitPoints ) {
            this.hitPoints = 0;
            return;
        }
        this.hitPoints -= dmg;
    }

    public void attack(Unit enemy) throws UnitIsDead {
        enemy.takeDamage(this.damage);
        System.out.println(this.name + " attacked " + enemy.name + "\n");

        enemy.ensureIsAlive();

        enemy.counterAttack(this);
    }

    public void counterAttack(Unit enemy) throws UnitIsDead {
        enemy.takeDamage(this.damage / 2);
        System.out.println(this.name + " counterattacked " + enemy.name + "\n");
    }

    public String toString() {
        if ( this.getHitPoints() == 0 ) {
            return  String.format(this.getName() + " is dead!");
        }
        return String.format("Unit: " + this.getName() + "\n"
                            + "Health: " + this.getHitPoints() + "/" + this.getHitPointsLimit() + "HP\n"
                            + "Damage: " + this.getDamage() + "dmg\n");
    }
}
